### PR TITLE
Composer Pipeline with Vertex Training, Model Upload and Deployment

### PR DESCRIPTION
**Pull request by kylesteckler**
The actual lab/example code is composer_vertex_lab.ipynb

The lab does the following:
1) Creates train/eval data tables in BigQuery for babyweight dataset
2) builds/pushes custom Vertex operators as Docker containers to the projects Container Registry for Vertex Training, Vertex Model Upload, and Vertex Model Deployment
3) Packages up a TF training application as a source distribution compatible with Vertex training. 
4) Creates a DAG file for a Composer pipeline that extracts data from BQ to CSVs -> Vertex Training -> Vertex Model Upload -> Vertex Deployment (using the custom operators). 
5) Uploads to DAG file to Composer and runs the pipeline.

Note: No lab and solution version yet, but I think this is a nice walkthrough/sample code of building a Composer pipeline that leverages Vertex services.

### CODE REVIEW

**Comment by BenoitDherin** [comment]
I don't think we should review this PR in this state. There are too many files and too much code for a single lab. If we are to merge this code I suggest to split this PR in much smaller PRs containing each at most one or two files with clear instructions regarding the necessary setup to test the code. 

I suggest we drop this PR entirely and draft a reasonable plan regarding composer inclusion so that we have a clear guide as how to refactor this code into something smaller and maintainable. 

**Minor note:** This PR contains commits that have failed the pre-commit check. Please run pre-commit.

**Comment by munnm** [comment]
I went through the solution notebook. It looks great! And everything works as expected. Thanks for pulling this together so quickly, Kyle.

There are a lot of files for a single lab; however, based on the conversations we've been having, I think having a lab that covers Composer and works for Vertex AI would be beneficial, even if not part of the core curriculum. A solution reference like this would be good to have in the repo for interested customers, either for future custom training or for open projects. 

To make the lab a bit more digestable, it might help to refactor into logical components; i.e. rather than write all BUILD config files, then build/push all images refactor so that you walkthrough each step separately before compiling into a pipeline. But that's just personal preference, it would depend a lot on the instruction slides. To Benoit's point, this may require a larger conversation as to whether to include Composer at all.

**Comment by BenoitDherin** [comment]
@munnm I am glad you were able to review and test the PR and that it works! and @kylesteckler Congrats for making this code work with such a quick turnaround! 

I am still concerned with length of this lab, the intricacy of the code, and having a repeatable setup of the infra required. We need something much smaller for the purpose of class and also something we can maintain if anything in the underlying platforms (either Vertex and/or Composer) changes.

Is there a way to simplify the logic in `vertex_train.py` and `vertex_deploy.py` for instance?
It seems to me there is a lot of unnecessary logic there for the context of class, and that the size of these file can be reduce by two or three. 

I am still maintaining we should drop this PR and, instead, breaking it into a sequence of several smaller independent PRs with less code in each, so that we can review and refactor each of them to get a more manageable version of this lab. Even if the overall code is working, I still don't think the code of this PR is in a suitable state for an effective lab that we can maintain over time. I am reluctant to put comments on the code with so many files, because it may takes us a while to merge the full PR with all these files if we proceed that way. 

As for breaking the PRs into more manageable ones, there is a natural way to do that:

* PR 1: commit only the code for training
* PR 2: add the code for model upload
* PR 3: add the code for model deploy

@kylesteckler : To pass the pre-commit checks, you can run `make precommit` at the repo top-level directory. It will try to fix the formatting of the code when it can, and give you pointers to the faulty lines when it can not. (You'll need to run `make install` first to have pre-commit setup on your node.)

File: notebooks/composer/composer_vertex/vertex_train_docker/vertex_train.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        state = client.get_custom_job(name=response.name).state
+        if state not in running_states:
+            break
+        time.sleep(arguments["job_polling_frequency"])
+    assert state == completed_state, "Job did not complete successfully."
+
+
+if __name__ == "__main__":
+    # Parse command line arguments.
+    arguments = parse_command_line_arguments()
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_train_docker/vertex_train.py:231]: Maybe replace with `fire` and remove the argument parsing logic?

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Use the official Python image.
+# https://hub.docker.com/_/python
+FROM python:3.8
+
+# Install python client.
+RUN pip install google-api-python-client
+RUN pip install --upgrade google-cloud-aiplatform
+
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/Dockerfile:8]: Consider pinning the libraries and avoiding `--upgrade``. (See comment above.)

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/build_image.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+gcloud builds submit --config=cloudbuild.json .
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/build_image.sh:3]: Same comments: `cloudbuild.json` -> `cloudbuild.yaml`; put instruction in the notebook & rm file.

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+import argparse
+
+from google.cloud import aiplatform
+
+
+def parse_arguments(parser):
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py:6]: Use fire instead.

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    )
+    parser.add_argument(
+        "--custom_serving_container_predict_route",
+        help="HTTP path on the container to send prediction requests to.",
+        type=str,
+        default="/predict"
+    )
+
+
+def parse_command_line_arguments():
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py:62]: Use `fire` instead

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    return arguments
+
+
+def upload_model(arguments):
+    vertex_native_ml_frameworks = set(
+        ["tensorflow", "pytorch", "xgboost", "sklearn"]
+    )
+    if arguments["ml_framework"] not in vertex_native_ml_frameworks:
+        prebuilt_prefixes = set(
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py:84]: top-level constant

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    )
+    if model_name_match:
+        print(
+            "Model with name {} already exists!".format(
+                model_display_name
+            )
+        )
+        model_display_name = "{}_0".format(model_display_name)
+
+    if arguments["ml_framework"] in vertex_native_ml_frameworks:
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py:116]: Maybe simplify this logic to allow a single framework.

File: notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            serving_container_predict_route=arguments["custom_serving_container_predict_route"],
+            serving_container_health_route=arguments["custom_serving_container_health_route"],
+            artifact_uri=arguments["artifact_uri"]
+        )
+    model_id = response.name.split("/")[-1]
+    print("Model ID = {}".format(model_id))
+
+
+if __name__ == "__main__":
+    arguments = parse_command_line_arguments()
```
Comment (BenoitDherin) [inline notebooks/composer/composer_vertex/vertex_upload_model_docker/vertex_upload_model.py:135]: Use `fire`.

**Comment by BenoitDherin** [comment]
I added comment and ideas around how to shrink this codebase, along with other minor issues.  I'll review the notebook next.

**Comment by kylesteckler** [comment]
Closing PR without merge. Will break this code apart and move forward with all the great feedback to simplify Composer/Vertex as a lab.

**Pull request closed without merging**